BugFix - Incorrect calculation of Error % for ColumnBasedChecks

samratmitra-0812 opened the pull request:
This PR fixes the issue of incorrect calculation of Error % for `ColumnBasedChecks`.

The result of a `RowBased` check can be completely defined in terms of total rows, error rows, and % of error rows w.r.t total rows. The message shown in the HTML and JSON reports can also be based on a common template for RowBased checks. This is being done in `ValidatorCheckEvent`.

But the above does not hold true for `ColumnBased` checks.

1) For `ColumnBased` checks, the calculation of Error % should be done based on the expectations specified in the yaml conf file, and the actual value calculated by the validator. The count of rows should not be a part of this calculation.

2) The fields needed to completely describe the result of a `ColumnBased` check may vary from one check to another. Even for the same check, it may vary with the data type of the column that we are validating. For example, within `ColumnMaxCheck`, the Error % field does not make sense for a `String` column, but is required for `Numeric` columns. This also implies that the message used in the reports to describe the results of a `ColumnBased` check will vary from one check to another.

This changes in this PR were made with the above points in mind. Some unit tests have also been added.

colindean commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -62,41 +74,66 @@ case class ColumnMaxCheck(column: String, value: Json)
+            actual = row.getDouble(idx)
+        }
+        pctError = if(expected != actual) calculatePctError(expected, actual) else "0.00"
+        data.appendAll(List(("Expected", num.toString), ("Actual", rMax.toString), ("Error Pct", pctError)))
```
```suggestion
        data.appendAll(List(("expected", num.toString), ("actual", rMax.toString), ("error_percent", pctError)))
```

samratmitra-0812 replied:
Done

colindean commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -10,11 +10,20 @@ import org.apache.spark.sql.catalyst.expressions.Expression
+
+  // calculates and returns the pct error as a string
+  def calculatePctError(expected: Double, actual: Double, formatStr: String = "%4.2f%%"): String = {
+    val pct = abs(((expected - actual) * 100.0) / expected)
```
There are a few places where you're doing something to guard against div/0. I think you should move that guard here.

samratmitra-0812 replied:
The guards that I was using was to check for validation status (pass/fail), and not for div/0. Hence I have not removed them.
 But thanks for catching the div/0 issue. Now I have handled the issue in the method `calculatePctError`.

colindean commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -62,41 +74,66 @@ case class ColumnMaxCheck(column: String, value: Json)
+        expected != actual
+      }
+      case ut => {
+        logger.error(s"ColumnMaxCheck for type: $ut, Row: $row not implemented! Please file this as a bug.")
```
```suggestion
        logger.error(s"ColumnMaxCheck for type: $ut, Row: $row not implemented! Please open a bug report on the data-validator issue tracker.")
```

samratmitra-0812 replied:
Done

samratmitra-0812 commented on src/main/scala/com/target/data_validator/JsonEncoders.scala:
```diff
@@ -47,6 +49,14 @@ object JsonEncoders extends LazyLogging {
         ("errorCount", Json.fromLong(vce.errorCount))
       )
+      case cbvce: ColumnBasedValidatorCheckEvent =>
+        val fields = new ListBuffer[Tuple2[String, Json]]
```
I have switched from ListBuffer to LinkedHashMap.

samratmitra-0812 commented on src/main/scala/com/target/data_validator/JsonEncoders.scala:
```diff
@@ -47,6 +49,14 @@ object JsonEncoders extends LazyLogging {
+        fields.append(("type", Json.fromString("columnBasedCheckEvent")))
+        fields.append(("failed", Json.fromBoolean(cbvce.failed)))
+        fields.append(("message", Json.fromString(cbvce.msg)))
+        cbvce.data.foreach(x => fields.append((x._1, Json.fromString(x._2))))
```
Done

samratmitra-0812 commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -36,8 +45,11 @@ case class MinNumRows(minNumRows: Long) extends ColumnBased("", ValidatorBase.L0
     addEvent(ValidatorCounter("rowCount", count))
-    addEvent(ValidatorCheckEvent(failed, s"MinNumRowCheck $minNumRows ", count, 1))
+    val msg = s"MinNumRowsCheck Expected: ${minNumRows} Actual: ${count} Error %: ${pctError}"
+    val data = List(("Expected", minNumRows.toString), ("Actual", count.toString), ("Error Pct", pctError))
```
Done

samratmitra-0812 commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -62,41 +74,66 @@ case class ColumnMaxCheck(column: String, value: Json)
+
+    var pctError = "0.00"
+    var errorMsg = ""
+    val data = new ListBuffer[Tuple2[String, String]]
```
Switched to LinkedHashMap.

samratmitra-0812 commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -62,41 +74,66 @@ case class ColumnMaxCheck(column: String, value: Json)
-      case DoubleType => num.forall(_.toDouble != row.getDouble(idx))
-      case ut =>
-        logger.error(s"quickCheck for type: $ut, Row: $row not Implemented! Please file this as a bug.")
+      case StringType => {
```
Done. Not sure if what I did is the same as exactly what you had in mind. Let me know.

samratmitra-0812 commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -62,41 +74,66 @@ case class ColumnMaxCheck(column: String, value: Json)
+      case StringType => {
+        val expected = value.asString.getOrElse("")
+        val actual = row.getString(idx)
+        data.appendAll(List(("Expected", expected), ("Actual", actual)))
```
Done

samratmitra-0812 commented:
@colindean Thanks for reviewing this.

I guess I have addressed all your comments. I have switched from List/ListBuffer to Map/LinkedHashMap as I wanted to make sure that the fields are logged in the reports in the same order in which they were inserted.

samratmitra-0812 commented:
@phpisciuneri Please share your thoughts on this PR.

phpisciuneri commented:
See #41

samratmitra-0812 commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -10,11 +10,32 @@ import org.apache.spark.sql.catalyst.expressions.Expression
+      pct_error_str = formatStr.format(pct)
+    }
+
+    pct_error_str
```
Done

samratmitra-0812 commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -36,8 +57,11 @@ case class MinNumRows(minNumRows: Long) extends ColumnBased("", ValidatorBase.L0
     addEvent(ValidatorCounter("rowCount", count))
-    addEvent(ValidatorCheckEvent(failed, s"MinNumRowCheck $minNumRows ", count, 1))
+    val msg = s"MinNumRowsCheck Expected: ${minNumRows} Actual: ${count} Error %: ${pctError}"
+    val data = LinkedHashMap("expected" -> minNumRows.toString, "actual" -> count.toString, "error_percent" -> pctError)
+    addEvent(ColumnBasedValidatorCheckEvent(failed, data.toMap, msg))
```
@colindean I am using `LinkedHashMap` to ensure that the retrieval order is the same as the insertion order. Not that it is absolutely necessary, but I wanted these fields to be written to the json report file in the same order (expected, actual, error_percent), instead of some random different order for each check.

colindean replied:
That's a good enough reason for me!

samratmitra-0812 commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -62,41 +86,61 @@ case class ColumnMaxCheck(column: String, value: Json)
-        true // Fail check!
+
+    var errorMsg = ""
+    val data = new LinkedHashMap[String, String]
```
Same comment as above. However, I am now using the `empty` method to create the empty map.

samratmitra-0812 commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -62,41 +86,61 @@ case class ColumnMaxCheck(column: String, value: Json)
+    var errorMsg = ""
+    val data = new LinkedHashMap[String, String]
+
+    val resultForString = () => {
```
Done.

phpisciuneri commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -62,41 +83,61 @@ case class ColumnMaxCheck(column: String, value: Json)
-        true // Fail check!
+
+    var errorMsg = ""
+    val data = LinkedHashMap.empty[String, String]
```
This feels a bit heavy handed to me, for ultimately what is just a couple of entries.  Maybe `ListMap` instead?  

https://www.scala-lang.org/api/current/scala/collection/immutable/ListMap.html

phpisciuneri commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -62,41 +83,61 @@ case class ColumnMaxCheck(column: String, value: Json)
+
+      failed = expected != actual
+      data += ("expected" -> expected, "actual" -> actual)
+      errorMsg = s"ColumnMaxCheck $column[StringType]: Expected: $expected, Actual: $actual"
```
Maybe this ought to be the format for when the relative error is `undefined`?

samratmitra-0812 replied:
I thinks for the same (check, dataType) combination, the message format should be same. Hence I did not change this.

phpisciuneri commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -62,41 +83,61 @@ case class ColumnMaxCheck(column: String, value: Json)
+      failed = cmp_params._1 != cmp_params._2
+      val pctError = if(failed) calculatePctError(cmp_params._1, cmp_params._2) else "0.00%"
+      data += ("expected" -> num.toString, "actual" -> rMax.toString, "error_percent" -> pctError)
+      errorMsg = s"ColumnMaxCheck $column[$dataType]: Expected: $num, Actual: $rMax. Error %: ${pctError}"
```
Again, the redundant percentage symbol and let's be specific that this is relative error.  Also, this is basically the duplicate format for `MinNumRowsCheck` with some slight differences (the period and comma here).  This can be defined once and then used both cases.

```suggestion
      errorMsg = s"ColumnMaxCheck $column[$dataType]: Expected: $num, Actual: $rMax Relative Error: ${pctError}"
```

phpisciuneri commented on src/test/scala/com/target/data_validator/validator/ColumnBasedSpec.scala:
```diff
@@ -79,10 +87,49 @@ class ColumnBasedSpec extends FunSpec with Matchers with TestingSparkSession {
+      assert(columnMaxCheck.getEvents contains
+        ColumnBasedValidatorCheckEvent(true,
+          LinkedHashMap("expected" -> "100", "actual" -> "3", "error_percent" -> "97.00%").toMap,
+          "ColumnMaxCheck number[IntegerType]: Expected: 100, Actual: 3. Error %: 97.00%"))
```
If you incorporate my suggestions above the formatting here and below will need updated.

samratmitra-0812 commented:
@phpisciuneri Thanks for reviewing this.

For columnSumCheck, I think another question that needs to be answered is how to calculate the relative error in case of non-inclusive intervals. For example, if the interval is (0.0, 1.0) and the actual value is 1.1, then the relative error should be calculated w.r.t to what? 0.9? 0.99? Please share your thoughts.

I also saw that the `columSumCheck` is not supported for Byte type, but the `columnMaxCheck` does support it. Any specific reasons for this?

colindean commented:
> I also saw that the columSumCheck is not supported for Byte type, but the columnMaxCheck does support it. Any specific reasons for this?

An oversight! I've taken care of it in #45.

phpisciuneri commented:
@samratmitra-0812 

> For columnSumCheck, I think another question that needs to be answered is how to calculate the relative error in case of non-inclusive intervals. For example, if the interval is (0.0, 1.0) and the actual value is 1.1, then the relative error should be calculated w.r.t to what? 0.9? 0.99? Please share your thoughts.

I think for the example you shared it can just be calculated with respect to the upper bound so 10%.

The more troublesome scenario is when the actual value is the same value as a boundary value in the non-inclusive case.  Then the relative error would be 0, but the test has failed.  One way to work around this is how you suggest.  But the implementation seems like it would be messy and require a bit of work.  You would probably need to choose the precision that you increment or decrement based on the type of the result and value and the least significant digit in either.

Maybe for the case of non-inclusive and actual value is the same as a boundary value we just treat the error as undefined.  What do you think of that?

samratmitra-0812 commented:
@phpisciuneri I think we could go with your suggestion of 'undefined' relative error, but we must document it. I will put that in README.

samratmitra-0812 commented:
@phpisciuneri I have added the fix for `columnSumCheck`, and addressed other comments. Please take a look.

And since I had to change the same files, I have included @colindean 's #45 changes in this PR.

phpisciuneri commented:
@samratmitra-0812 

> And since I had to change the same files, I have included @colindean 's #45 changes in this PR.

Please do not copy code from another PR into yours.  It is not a good practice and we lose the history of who originally wrote the code.  I am going to go ahead and remove the copied code from this branch.

Instead please provide your review on #45.  See [About Pull Request Reviews](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-reviews).  If you approve, we will merge that branch into master, and then we can update this PR accordingly & address any merge conflicts at that point.  See [Addressing Merge Conflicts](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/addressing-merge-conflicts).

There are occasions where it may be appropriate to copy commits from elsewhere.  In that case use [cherry-picking](https://git-scm.com/docs/git-cherry-pick).

phpisciuneri commented on README.md:
```diff
@@ -324,6 +324,8 @@ This check sums a column in all rows. If the sum applied to the `column` doesn't
 | `maxValue`  | NumericType | The upper bound of the sum. Type depends on the type of the `column`.  |
 | `inclusive` | Boolean     | Include `minValue` and `maxValue` as part of the range.                |
 
+**Note:** If bounds are non-inclusive, and the actual sum is eequal to one of the bounds, the relative error percentage will be undefined.
```
```suggestion
**Note:** If bounds are non-inclusive, and the actual sum is equal to one of the bounds, the relative error percentage will be undefined.
```

samratmitra-0812 commented:
@phpisciuneri removed #45 changes from this PR.

phpisciuneri commented:
OK, we are getting closer.  Thanks for being patient.  One problem that I have now is that the structure of the checks in the report now contains a lot of duplication of the parameters.  See the snippet below.

FWIW: I used [jq](https://stedolan.github.io/jq/manual/#Basicfilters) to parse the report.  In my case, the report is named `report.json`:

```bash
jq .tables[0].checks report.json
```

```json
[
  {
    "type": "rowCount",
    "minNumRows": 50000,
    "failed": true,
    "events": [
      {
        "type": "counter",
        "name": "rowCount",
        "value": 31363
      },
      {
        "type": "columnBasedCheckEvent",
        "failed": true,
        "message": "MinNumRowsCheck Expected: 50000 Actual: 31363 Relative Error: 37.27%",
        "data": {
          "relative_error": "37.27%",
          "expected": "50000",
          "actual": "31363"
        }
      }
    ]
  },
  {
    "type": "columnMaxCheck",
    "column": "age",
    "value": 0,
    "failed": true,
    "events": [
      {
        "type": "columnBasedCheckEvent",
        "failed": true,
        "message": "ColumnMaxCheck age[IntegerType]: Expected: 0, Actual: 90. Relative Error: undefined",
        "data": {
          "relative_error": "undefined",
          "expected": "0",
          "actual": "90"
        }
      }
    ]
  },
  {
    "type": "columnSumCheck",
    "failed": true,
    "events": [
      {
        "type": "columnBasedCheckEvent",
        "failed": true,
        "message": "columnSumCheck on age[LongType]: Expected Range: (0 , 1e4) Actual: 1200747 Relative Error: 11907.47%",
        "data": {
          "inclusive": "false",
          "upper_bound": "1e4",
          "relative_error": "11907.47%",
          "actual": "1200747",
          "lower_bound": "0"
        }
      }
    ],
    "column": "age",
    "minValue": 0,
    "maxValue": 10000,
    "inclusive": null
  }
]
```

@samratmitra-0812 was there a particular reason for having all the error related terms inside of the `data` object?

Can/Should we simplify here or make that an issue going forward?

samratmitra-0812 commented:
@phpisciuneri Initially, this additional nesting for data was not there. @colindean suggested that we put the data in a separate object of its own as this would make the partial function consistent with the others.

> This should follow suit from the case below and put the data in a property of its own. This would make this partial function cleaner and follow the pattern used in the others in this method.

I guess we will need more time to refactor the structure of the json output. If I am not wrong, the duplication issue exists for the RowBased checks as well. We will need to decide what information we show in the report, and how this information is structured. 

We may also need to think about making the json output a little less verbose, and show only information that is actually useful. For example, it it necessary to write `ValidationTimer` events to the report?

I will create an issue for this.

colindean commented:
I'm OK with the duplication of parameters in each check's output. It's sensible for each check to report how it was configured. It's also sensible for the check's events to have a human-readable message as well as the data used in that message as a queriable object. I'd drop the human-readable message before the error object.

phpisciuneri commented on src/test/scala/com/target/data_validator/ValidatorBaseSpec.scala:
```diff
@@ -10,6 +10,7 @@ import org.apache.spark.sql.catalyst.expressions.Literal
 import org.apache.spark.sql.types._
 import org.scalatest._
 
+import scala.collection.mutable.ListMap
```
```suggestion
import scala.collection.immutable.ListMap
```

phpisciuneri commented on src/test/scala/com/target/data_validator/ValidatorBaseSpec.scala:
```diff
@@ -182,20 +183,28 @@ class ValidatorBaseSpec extends FunSpec with Matchers with TestingSparkSession {
       assert(config.quickChecks(spark, dict))
       assert(config.failed)
       assert(config.tables.head.failed)
+      assert(minNumRowsCheck.getEvents contains ColumnBasedValidatorCheckEvent(true,
+        ListMap("expected" -> "10", "actual" -> "2", "relative_error" -> "80.00%").toMap,
+        "MinNumRowsCheck Expected: 10 Actual: 2 Relative Error: 80.00%"))
```
```suggestion
      assert(minNumRowsCheck.getEvents contains ColumnBasedValidatorCheckEvent(
        failure = true,
        ListMap("expected" -> "10", "actual" -> "2", "relative_error" -> "80.00%"),
        "MinNumRowsCheck Expected: 10 Actual: 2 Relative Error: 80.00%"
      ))
```

phpisciuneri commented on src/test/scala/com/target/data_validator/ValidatorBaseSpec.scala:
```diff
@@ -182,20 +183,28 @@ class ValidatorBaseSpec extends FunSpec with Matchers with TestingSparkSession {
       assert(!config.quickChecks(spark, dict))
       assert(!config.failed)
       assert(!config.tables.exists(_.failed))
+      assert(minNumRowsCheck.getEvents contains ColumnBasedValidatorCheckEvent(false,
+        ListMap("expected" -> "1", "actual" -> "2", "relative_error" -> "0.00%").toMap,
+        "MinNumRowsCheck Expected: 1 Actual: 2 Relative Error: 0.00%"))
```
```suggestion
      assert(minNumRowsCheck.getEvents contains ColumnBasedValidatorCheckEvent(
        failure = false,
        ListMap("expected" -> "1", "actual" -> "2", "relative_error" -> "0.00%"),
        "MinNumRowsCheck Expected: 1 Actual: 2 Relative Error: 0.00%"
      ))
```

phpisciuneri commented on src/test/scala/com/target/data_validator/ValidatorBaseSpec.scala:
```diff
@@ -182,20 +183,28 @@ class ValidatorBaseSpec extends FunSpec with Matchers with TestingSparkSession {
       val df = spark.createDataFrame(sc.parallelize(List(Row("Doug", 50), Row("Collin", 32))), schema) //scalastyle:ignore
-      val config = mkConfig(df, List(MinNumRows(1))) //scalastyle:ignore
+      val minNumRowsCheck = MinNumRows(1)
+      val config = mkConfig(df, List(minNumRowsCheck)) //scalastyle:ignore
```
```suggestion
      val config = mkConfig(df, List(minNumRowsCheck))
```

phpisciuneri commented on src/test/scala/com/target/data_validator/ValidatorBaseSpec.scala:
```diff
@@ -182,20 +183,28 @@ class ValidatorBaseSpec extends FunSpec with Matchers with TestingSparkSession {
       val dict = new VarSubstitution
       val df = spark.createDataFrame(sc.parallelize(List(Row("Doug", 50), Row("Collin", 32))), schema) //scalastyle:ignore
-      val config = mkConfig(df, List(MinNumRows(10))) //scalastyle:ignore
+      val minNumRowsCheck = MinNumRows(10)
+      val config = mkConfig(df, List(minNumRowsCheck)) //scalastyle:ignore
```
```suggestion
      val minNumRowsCheck = MinNumRows(10) // scalastyle:ignore magic.number
      val config = mkConfig(df, List(minNumRowsCheck))
```

phpisciuneri commented on src/test/scala/com/target/data_validator/validator/ColumnBasedSpec.scala:
```diff
@@ -8,6 +8,8 @@ import org.apache.spark.sql.Row
 import org.apache.spark.sql.types._
 import org.scalatest._
 
+import scala.collection.mutable.ListMap
```
```suggestion
import scala.collection.immutable.ListMap
```

phpisciuneri commented on src/test/scala/com/target/data_validator/validator/ColumnBasedSpec.scala:
```diff
@@ -63,10 +66,15 @@ class ColumnBasedSpec extends FunSpec with Matchers with TestingSparkSession {
       assert(!sut.configCheck(spark, dict))
       assert(sut.quickChecks(spark, dict))
       assert(sut.failed)
+      assert(columnMaxCheck.getEvents contains
+        ColumnBasedValidatorCheckEvent(true,
+          ListMap("expected" -> "2018/11/01", "actual" -> "2018/10/31").toMap,
+          "ColumnMaxCheck data[StringType]: Expected: 2018/11/01, Actual: 2018/10/31"))
```
```suggestion
      assert(columnMaxCheck.getEvents contains ColumnBasedValidatorCheckEvent(
        failure = true,
        ListMap("expected" -> "2018/11/01", "actual" -> "2018/10/31"),
        "ColumnMaxCheck data[StringType]: Expected: 2018/11/01 Actual: 2018/10/31"
      ))
```

phpisciuneri commented on src/test/scala/com/target/data_validator/validator/ColumnBasedSpec.scala:
```diff
@@ -79,10 +87,49 @@ class ColumnBasedSpec extends FunSpec with Matchers with TestingSparkSession {
     it("should fail when numeric column doesn't match max value") {
       val dict = new VarSubstitution
-      val sut = mkValidatorConfig(List(ColumnMaxCheck("number", Json.fromInt(100)))) // scalastyle:ignore
+      val columnMaxCheck = ColumnMaxCheck("number", Json.fromInt(100))
+      val sut = mkValidatorConfig(List(columnMaxCheck)) // scalastyle:ignore
```
```suggestion
      val columnMaxCheck = ColumnMaxCheck("number", Json.fromInt(100)) // scalastyle:ignore magic.number
      val sut = mkValidatorConfig(List(columnMaxCheck))
```

phpisciuneri commented on src/test/scala/com/target/data_validator/validator/ColumnBasedSpec.scala:
```diff
@@ -79,10 +87,49 @@ class ColumnBasedSpec extends FunSpec with Matchers with TestingSparkSession {
+      assert(!sut.configCheck(spark, dict))
+      assert(sut.quickChecks(spark, dict))
+      assert(sut.failed)
+      assert(columnMaxCheck.getEvents contains
+        ColumnBasedValidatorCheckEvent(true,
+          ListMap("expected" -> "100", "actual" -> "3", "relative_error" -> "97.00%").toMap,
+          "ColumnMaxCheck number[IntegerType]: Expected: 100, Actual: 3. Relative Error: 97.00%"))
```
```suggestion
      assert(columnMaxCheck.getEvents contains ColumnBasedValidatorCheckEvent(
        failure = true,
        ListMap("expected" -> "100", "actual" -> "3", "relative_error" -> "97.00%"),
        "ColumnMaxCheck number[IntegerType]: Expected: 100 Actual: 3 Relative Error: 97.00%"
      ))
```

phpisciuneri commented on src/test/scala/com/target/data_validator/validator/ColumnBasedSpec.scala:
```diff
@@ -79,10 +87,49 @@ class ColumnBasedSpec extends FunSpec with Matchers with TestingSparkSession {
+      assert(!sut.configCheck(spark, dict))
+      assert(sut.quickChecks(spark, dict))
+      assert(sut.failed)
+      assert(columnMaxCheck.getEvents contains
+        ColumnBasedValidatorCheckEvent(true,
+          ListMap("expected" -> "0", "actual" -> "3", "relative_error" -> "undefined").toMap,
+          "ColumnMaxCheck number[IntegerType]: Expected: 0, Actual: 3. Relative Error: undefined"))
```
```suggestion
      assert(columnMaxCheck.getEvents contains ColumnBasedValidatorCheckEvent(
        failure = true,
        ListMap("expected" -> "0", "actual" -> "3", "relative_error" -> "undefined"),
        "ColumnMaxCheck number[IntegerType]: Expected: 0 Actual: 3 Relative Error: undefined"
      ))
```

phpisciuneri commented on src/test/scala/com/target/data_validator/validator/ColumnBasedSpec.scala:
```diff
@@ -79,10 +87,49 @@ class ColumnBasedSpec extends FunSpec with Matchers with TestingSparkSession {
+    it("should fail with undefined error % when numeric column doesn't match max value and expected value is 0") {
+      val dict = new VarSubstitution
+      val columnMaxCheck = ColumnMaxCheck("number", Json.fromInt(0))
+      val sut = mkValidatorConfig(List(columnMaxCheck)) // scalastyle:ignore
```
```suggestion
      val sut = mkValidatorConfig(List(columnMaxCheck))
```

phpisciuneri commented on src/test/scala/com/target/data_validator/validator/ColumnBasedSpec.scala:
```diff
@@ -79,10 +87,49 @@ class ColumnBasedSpec extends FunSpec with Matchers with TestingSparkSession {
       assert(!sut.configCheck(spark, dict))
       assert(sut.quickChecks(spark, dict))
       assert(sut.failed)
+      assert(columnMaxCheck.getEvents contains
+        ColumnBasedValidatorCheckEvent(true,
+          ListMap("expected" -> "5.0", "actual" -> "3.5", "relative_error" -> "30.00%").toMap,
+          "ColumnMaxCheck double[DoubleType]: Expected: 5.0, Actual: 3.5. Relative Error: 30.00%"))
```
```suggestion
      assert(columnMaxCheck.getEvents contains ColumnBasedValidatorCheckEvent(
        failure = true,
        ListMap("expected" -> "5.0", "actual" -> "3.5", "relative_error" -> "30.00%"),
        "ColumnMaxCheck double[DoubleType]: Expected: 5.0 Actual: 3.5 Relative Error: 30.00%"
      ))
```

phpisciuneri commented on src/test/scala/com/target/data_validator/validator/ColumnBasedSpec.scala:
```diff
@@ -79,10 +87,49 @@ class ColumnBasedSpec extends FunSpec with Matchers with TestingSparkSession {
+    it("should fail when double column doesn't match max value") {
+      val dict = new VarSubstitution
+      val columnMaxCheck = ColumnMaxCheck("double", Json.fromDouble(5.0).get)
+      val sut = mkValidatorConfig(List(columnMaxCheck)) // scalastyle:ignore
```
```suggestion
      val sut = mkValidatorConfig(List(columnMaxCheck))
```

phpisciuneri commented on src/main/scala/com/target/data_validator/validator/ColumnBased.scala:
```diff
@@ -36,8 +54,12 @@ case class MinNumRows(minNumRows: Long) extends ColumnBased("", ValidatorBase.L0
     addEvent(ValidatorCounter("rowCount", count))
-    addEvent(ValidatorCheckEvent(failed, s"MinNumRowCheck $minNumRows ", count, 1))
+    val msg = s"MinNumRowsCheck Expected: $minNumRows Actual: $count Relative Error: $pctError"
+    val data = ListMap("expected" -> minNumRows.toString, "actual" -> count.toString,
+                             "relative_error" -> pctError)
```
```suggestion
    val data = ListMap("expected" -> minNumRows.toString, "actual" -> count.toString, "relative_error" -> pctError)
```

phpisciuneri commented on src/main/scala/com/target/data_validator/validator/ColumnSumCheck.scala:
```diff
@@ -52,28 +55,66 @@ case class ColumnSumCheck(
+      }
+    }
+
+    def getData(actualSum: Double, pctError: String): ListMap[String, String] = {
+      val data = (minValue, maxValue) match {
```
```suggestion
    def getData(pctError: String): ListMap[String, String] = {
      ((minValue, maxValue) match {
```

phpisciuneri commented on src/main/scala/com/target/data_validator/validator/ColumnSumCheck.scala:
```diff
@@ -52,28 +55,66 @@ case class ColumnSumCheck(
+        case (None, Some(y)) => ListMap("upper_bound" -> y.asNumber.get.toString)
+        case (Some(x), None) => ListMap("lower_bound" -> x.asNumber.get.toString)
+        case (None, None) => throw new RuntimeException("Must define at least one of minValue or maxValue.")
+      }
+
+      data += ("inclusive" -> isInclusive.toString, "actual" -> r(idx).toString, "relative_error" -> pctError)
```
```suggestion
      }) + ("inclusive" -> isInclusive.toString, "actual" -> r(idx).toString, "relative_error" -> pctError)
```
